contents/ruby/Rails.md

```diff
@@ -0,0 +1,170 @@
+
+The structure of an app written in the MVC framework is as follows:
+
+<center>
+<img src="mvc.jpeg">
+
+_Figure 1. Model-View-Controller Framework_ <sup>[source](https://medium.com/@matthewmain/rails-request-response-cycle-819e9cd8fa4e)</sup>
+</center>
+
+Let us follow the diagram, and trace the control flow as a Rails app responds to a request by the client. In this example, the client is making a query for all users in the database. 
+
+#### Router
+
+When Rails receives the URL, it first looks up the 
+<tooltip effect="scale" content="A URI (Uniform Resource Identifier) is a string of characters that are used to identify a resource. URLs are a subset of URIs.">
+<a href="">URI</a>
+</tooltip> in the `routes.rb` file, which defines, by default, all the routes in the application. Each valid URI is mapped to a function located inside a Controller, that is then invoked to provide a response to the request. 
+
+In order to achieve this, Rails exposes a convenient abstraction called a "Resource". A resource can refer to an object upon which you can perform CRUD (create, read, update, and delete) operations. These operations are often invoked via a HTTP request of the appropriate request verb (POST, GET, PUT, and DELETE respectively). In order to declare a resource called `user`, you would mention it inside your `routes.rb` in the following manner: 
+
+```ruby
+resources :users
+```
+
+Now, you can check all the routes that you have in your application by running `rails routes` in the command line. This should include:
+
+```
+                   Prefix Verb   URI Pattern                                                                              Controller#Action
+                    users GET    /users(.:format)                                                                         users#index
+                          POST   /users(.:format)                                                                         users#create
+                 new_user GET    /users/new(.:format)                                                                     users#new
+                edit_user GET    /users/:id/edit(.:format)                                                                users#edit
+                     user GET    /users/:id(.:format)                                                                     users#show
+                          PATCH  /users/:id(.:format)                                                                     users#update
+                          PUT    /users/:id(.:format)                                                                     users#update
+                          DELETE /users/:id(.:format)                                                                     users#destroy
+```
```

Comment:
I just noticed that there's a large chunk of content that's been "missing" all this while, because you'd have to scroll so much to the right to realize that it's there:

<img width="885" alt="Screenshot 2019-04-17 at 1 29 54 AM" src="https://user-images.githubusercontent.com/9073504/56231170-6108fb80-60b0-11e9-83f0-2592551667ab.png">

You should reduce the number of spaces between the columns so that they show up nicely in the website.

Also, remove the excess whitespace padding to the left of the `Prefix` column.

Reply:
Ah dear I just copied that from my terminal. Okay gotcha :)